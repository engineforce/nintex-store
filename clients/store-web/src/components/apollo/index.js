import React, { Component } from 'react'
import ApolloClient, { InMemoryCache, gql } from 'apollo-boost'
import { persistCache } from 'apollo-cache-persist'
import fetch from 'isomorphic-fetch'
import { ApolloProvider } from 'react-apollo' // Or just react-apollo

const cache = new InMemoryCache()

const typeDefs = gql`
  extend type Query {
    orderItems: [OrderItem]!
  }
  type OrderItem {
    productId: String!
    quantity: Int!
  }
`

const client = new ApolloClient({
  cache,
  fetch,
  // Paul Debug: test only
  uri: 'https://countries.trevorblades.com/',
  typeDefs,
})

export default class extends Component {
  state = {
    client: null,
    loaded: false,
  }

  async componentDidMount() {
    await persistCache({
      cache,
      storage: window.localStorage,
      maxSize: false,
      debug: process.env.NODE_ENV !== 'production',
    })

    this.setState({
      client,
    })
  }

  render() {
    const { client } = this.state
    const { children } = this.props
    if (!client) return <div>Loading...</div>
    return <ApolloProvider client={client}>{children}</ApolloProvider>
  }
}
