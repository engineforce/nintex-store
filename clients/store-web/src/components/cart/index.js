import React, { useState } from 'react'
import { useStaticQuery, graphql } from 'gatsby'
import ProductImage from '../product-image'
import {
  Root,
  Grid,
  Row,
  HeaderRow,
  Column,
  DescriptionColumn,
  PriceColumn,
  QuantityColumn,
  Product,
  ProductImageWrapper,
  Input,
} from './styled'
import gql from 'graphql-tag'
import { Query } from 'react-apollo'
import { pipe, reduce, map, prop } from 'ramda'
import { calculateOrderPrice } from 'price-calculators'

const formatCurrency = new Intl.NumberFormat('en-AU', {
  style: 'currency',
  currency: 'AUD',
}).format

const GET_ORDER_ITEMS = gql`
  {
    orderItems @client {
      productId
      quantity
    }
  }
`

const Cart = () => {
  const {
    allContentfulProduct: { edges },
  } = useStaticQuery(
    graphql`
      query {
        allContentfulProduct(filter: { node_locale: { eq: "en-US" } }) {
          edges {
            node {
              productId
              price
              name
            }
          }
        }
      }
    `
  )

  const products = getProductsMap(edges)

  const [promoCode, setPromoCode] = useState('')

  return (
    <Root>
      <Query query={GET_ORDER_ITEMS} fetchPolicy={'cache-only'}>
        {({ data, loading, error = undefined }) => {
          if (loading) return <p>Loading...</p>

          const orderItems = !error && data.orderItems ? data.orderItems : []

          return (
            <Grid>
              <HeaderRow>
                <DescriptionColumn>My Cart</DescriptionColumn>
                <PriceColumn>Price</PriceColumn>
                <QuantityColumn>Qty</QuantityColumn>
              </HeaderRow>
              <hr />
              {orderItems.map(({ productId, quantity }) => {
                const { name, price } = products[productId]
                return (
                  <Row key={productId}>
                    <DescriptionColumn>
                      <Product>
                        <ProductImageWrapper>
                          <ProductImage key={productId} productId={productId} />
                        </ProductImageWrapper>
                        {name}
                      </Product>
                    </DescriptionColumn>
                    <PriceColumn>{price}</PriceColumn>
                    {/* Paul Debug: quantity cannot be changed yet */}
                    <QuantityColumn>
                      <Input value={quantity}></Input>
                    </QuantityColumn>
                  </Row>
                )
              })}

              <hr />
              <Row>
                <Column>
                  <span>Promo Code</span>
                  <Input
                    value={promoCode}
                    onChange={event => setPromoCode(event.target.value)}
                  />
                </Column>
                <Column>
                  {renderTotal({
                    orderItems,
                    products,
                    promoCode,
                  })}
                </Column>
              </Row>
            </Grid>
          )
        }}
      </Query>
    </Root>
  )
}

function renderTotal({ orderItems, products, promoCode }) {
  const { total, discountedTotal } = calculateOrderPrice({
    orderItems,
    products,
    promoCode,
  })

  return (
    <>
      <span>Total</span>
      <div>{formatCurrency(total)}</div>

      {discountedTotal && discountedTotal !== total && (
        <>
          <span>Discounted Total</span>
          <div>{formatCurrency(discountedTotal)}</div>
        </>
      )}
    </>
  )
}

const getProductsMap = pipe(
  map(prop('node')),
  reduce((obj, item) => {
    obj[item['productId']] = item
    return obj
  }, {})
)

export default Cart
