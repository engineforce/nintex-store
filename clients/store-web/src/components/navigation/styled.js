import styled from 'styled-components'

export const Root = styled.div`
  display: flex;
  flex-direction: row;
  flex-wrap: nowrap;
  justify-content: center;
  align-content: center;
  align-items: flex-start;
`

export const NavItem = styled.div`
  width: 100px;
`
