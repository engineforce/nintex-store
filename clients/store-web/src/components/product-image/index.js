import React from 'react'
import { useStaticQuery, graphql } from 'gatsby'
import Img from 'gatsby-image'

const ProductImage = ({ productId }) => {
  const images = useStaticQuery(graphql`
    fragment ImageFile on File {
      childImageSharp {
        fluid(maxWidth: 600) {
          ...GatsbyImageSharpFluid
        }
      }
    }

    query {
      form: file(
        sourceInstanceName: { eq: "product-images" }
        relativePath: { eq: "form.png" }
      ) {
        ...ImageFile
      }
      wf: file(
        sourceInstanceName: { eq: "product-images" }
        relativePath: { eq: "workflow.png" }
      ) {
        ...ImageFile
      }
      docgen: file(
        sourceInstanceName: { eq: "product-images" }
        relativePath: { eq: "docgen.jpg" }
      ) {
        ...ImageFile
      }
    }
  `)

  return <Img fluid={images[productId].childImageSharp.fluid} />
}

export default ProductImage
