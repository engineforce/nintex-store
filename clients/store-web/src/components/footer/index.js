import React from 'react'
import { Root } from './styled'

const Footer = () => (
  <Root>
    <span>
      Built by {` `}{' '}
      <a target="blank" href="https://www.linkedin.com/in/paul-li-97381b3/">
        engineforce
      </a>
      , Â© {new Date().getFullYear()}
    </span>
  </Root>
)

export default Footer
