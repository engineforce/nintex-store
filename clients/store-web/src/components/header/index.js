import { Link } from 'gatsby'
import PropTypes from 'prop-types'
import React from 'react'
import { Root, H1 } from './styled'

const Header = ({ siteTitle }) => (
  <Root>
    <H1>
      <Link to="/">{siteTitle}</Link>
    </H1>
  </Root>
)

Header.propTypes = {
  siteTitle: PropTypes.string,
}

Header.defaultProps = {
  siteTitle: ``,
}

export default Header
