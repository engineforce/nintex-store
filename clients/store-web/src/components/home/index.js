import { Link } from 'gatsby'
import PropTypes from 'prop-types'
import React from 'react'
import { Root } from './styled'

const Home = () => (
  <Root>
    <h2>Welcome to Nintex Online Store.</h2>
    <p>
      Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod
      tempor incididunt ut labore et dolore magna aliqua. Aliquam etiam erat
      velit scelerisque in dictum. Vivamus at augue eget arcu dictum varius duis
      at consectetur. Malesuada pellentesque elit eget gravida cum. Pellentesque
      nec nam aliquam sem et tortor consequat id porta. Quam quisque id diam vel
      quam elementum pulvinar etiam. Massa massa ultricies mi quis hendrerit
      dolor. Quis hendrerit dolor magna eget est. Nec dui nunc mattis enim ut
      tellus elementum sagittis. At erat pellentesque adipiscing commodo.
      Imperdiet dui accumsan sit amet nulla facilisi morbi tempus iaculis. Elit
      eget gravida cum sociis. Phasellus egestas tellus rutrum tellus
      pellentesque eu. Nisl condimentum id venenatis a condimentum vitae sapien
      pellentesque habitant. At risus viverra adipiscing at in tellus integer.
    </p>
    <p>
      Diam maecenas ultricies mi eget mauris pharetra et ultrices neque. Vitae
      tortor condimentum lacinia quis. Aliquam ultrices sagittis orci a
      scelerisque purus. Velit ut tortor pretium viverra suspendisse potenti
      nullam ac. Est sit amet facilisis magna. Consectetur purus ut faucibus
      pulvinar elementum integer enim. Sed augue lacus viverra vitae congue eu
      consequat. Aliquam sem et tortor consequat id porta nibh venenatis cras.
      Congue nisi vitae suscipit tellus mauris a diam maecenas. Suspendisse
      interdum consectetur libero id faucibus nisl tincidunt eget nullam. Eget
      sit amet tellus cras adipiscing enim. Porta non pulvinar neque laoreet
      suspendisse. Ac tincidunt vitae semper quis lectus nulla at. Nisl pretium
      fusce id velit ut tortor pretium viverra suspendisse.
    </p>
  </Root>
)

Home.propTypes = {
  siteTitle: PropTypes.string,
}

Home.defaultProps = {
  siteTitle: ``,
}

export default Home
