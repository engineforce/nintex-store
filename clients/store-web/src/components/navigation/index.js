import React from 'react'
import { Link } from 'gatsby'
import { Root, NavItem } from './styled'
import gql from 'graphql-tag'
import { Query } from 'react-apollo'
import { reduce } from 'ramda'

export const GET_ORDER_ITEMS = gql`
  {
    orderItems @client {
      productId
      quantity
    }
  }
`

const Navigation = () => (
  <Root>
    <NavItem>
      <Link to="/">Home</Link>
    </NavItem>
    <NavItem>
      <Link to="/products">Products</Link>
    </NavItem>
    <NavItem>
      <Query query={GET_ORDER_ITEMS}>
        {({ data, loading, error = undefined }) => {
          return (
            <Link to="/cart">
              Cart (
              {!loading && !error && data.orderItems
                ? reduce(
                    (count, item) => count + item.quantity,
                    0,
                    data.orderItems
                  )
                : 0}
              )
            </Link>
          )
        }}
      </Query>
    </NavItem>
  </Root>
)

export default Navigation
