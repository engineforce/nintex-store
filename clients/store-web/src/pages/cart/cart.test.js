import React from 'react'
import { MockedProvider } from 'react-apollo/test-utils'
import renderer from 'react-test-renderer'
import CartPage from './index'
import { useStaticQuery } from 'gatsby'
import wait from 'waait'
import { GET_ORDER_ITEMS } from '../../components/navigation'
jest.mock('../../components/product-image', () => 'ProductImage')

describe('CartPage', () => {
  describe('when render with no item in cart', () => {
    const { create, mocks } = getTestContext({
      useStaticQuery,
    })

    beforeAll(async () => {
      mocks.useStaticQuery.mockImplementation(() => ({
        ...getSiteData(),
        ...getProductData(),
      }))

      create(<CartPage />)
      await wait(0)
    })

    it('', () => {
      /** @type {renderer.ReactTestRenderer} */
      const home = create.mock.results[0].value
      expect(home.toJSON()).toMatchSnapshot()
    })
  })

  describe('when render with 1 item in cart', () => {
    const { create, mocks } = getTestContext({
      useStaticQuery,
      apolloData: [
        {
          request: {
            query: GET_ORDER_ITEMS,
          },
          result: {
            data: {
              orderItems: [{ productId: 'form', quantity: 1 }],
            },
          },
        },
      ],
    })

    beforeAll(async () => {
      mocks.useStaticQuery.mockImplementation(() => ({
        ...getSiteData(),
        ...getProductData(),
      }))

      create(<CartPage />)
      await wait(0)
    })

    it('', () => {
      /** @type {renderer.ReactTestRenderer} */
      const home = create.mock.results[0].value
      expect(home.toJSON()).toMatchSnapshot()
    })
  })
})

function getTestContext({ useStaticQuery, apolloData = undefined }) {
  const mocks = {
    useStaticQuery,
    apollo: apolloData || [
      {
        request: {
          query: GET_ORDER_ITEMS,
        },
        result: {
          data: {
            orderItems: [],
          },
        },
      },
    ],
  }
  return {
    mocks,
    create: jest.fn().mockImplementation(element =>
      renderer.create(
        <MockedProvider addTypename={false} mocks={mocks.apollo}>
          {element}
        </MockedProvider>
      )
    ),
  }
}

function getSiteData() {
  return {
    site: {
      siteMetadata: {
        title: 'Online Store',
      },
    },
  }
}

function getProductData() {
  return {
    allContentfulProduct: {
      edges: [
        {
          node: {
            productId: 'wf',
            price: 199.99,
            description: {
              description: 'Workflow description',
            },
            name: 'Workflow',
          },
        },
        {
          node: {
            productId: 'form',
            price: 99.99,
            description: {
              description: 'Form description',
            },
            name: 'Form',
          },
        },
        {
          node: {
            productId: 'docgen',
            price: 9.99,
            description: {
              description: 'Document Generation description',
            },
            name: 'Document Generation',
          },
        },
      ],
    },
  }
}
