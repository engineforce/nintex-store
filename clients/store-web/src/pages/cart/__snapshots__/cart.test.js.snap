// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CartPage when render with 1 item in cart  1`] = `
Array [
  <header
    className="sc-bdVaJa llhFdL"
  >
    <h1
      className="sc-bwzfXH kabNnl"
    >
      <a
        aria-current="page"
        className=""
        href="/"
        onClick={[Function]}
        onMouseEnter={[Function]}
        style={Object {}}
      >
        Online Store
      </a>
    </h1>
  </header>,
  <nav
    className="sc-htpNat bRwrsF"
  >
    <div
      className="sc-bxivhb cLBQgY"
    >
      <a
        aria-current="page"
        className=""
        href="/"
        onClick={[Function]}
        onMouseEnter={[Function]}
        style={Object {}}
      >
        Home
      </a>
    </div>
    <div
      className="sc-bxivhb cLBQgY"
    >
      <a
        href="/products"
        onClick={[Function]}
        onMouseEnter={[Function]}
      >
        Products
      </a>
    </div>
    <div
      className="sc-bxivhb cLBQgY"
    >
      <a
        href="/cart"
        onClick={[Function]}
        onMouseEnter={[Function]}
      >
        Cart (
        1
        )
      </a>
    </div>
  </nav>,
  <main
    className="sc-EHOje ksOsdW"
  >
    <div
      className="sc-bZQynM iwmxdw"
    >
      <ul
        className="sc-gzVnrw lfdfns"
      >
        <li
          className="sc-htoDjs bPGBDH"
        >
          <div>
            <button
              className="sc-cSHVUG iFKOBr"
              onClick={[Function]}
            >
              EMPTY CART
            </button>
          </div>
        </li>
        <li
          className="sc-htoDjs sc-dnqmqq BIZtQ"
        >
          <div
            className="sc-iwsKbI sc-gZMcBi gdTgYj"
          >
            My Cart
          </div>
          <div
            className="sc-iwsKbI sc-gqjmRU fQHcrV"
          >
            Price
          </div>
          <div
            className="sc-iwsKbI sc-VigVT kGKNKu"
          >
            Qty
          </div>
        </li>
        <hr />
        <li
          className="sc-htoDjs bPGBDH"
        >
          <div
            className="sc-iwsKbI sc-gZMcBi gdTgYj"
          >
            <div
              className="sc-jTzLTM eoQMeB"
            >
              <div
                className="sc-fjdhpX jOjPCH"
              >
                <ProductImage
                  productId="form"
                />
              </div>
              Form
            </div>
          </div>
          <div
            className="sc-iwsKbI sc-gqjmRU fQHcrV"
          >
            A$99.99
          </div>
          <div
            className="sc-iwsKbI sc-VigVT kGKNKu"
          >
            <input
              className="sc-jzJRlG fsfCyH"
              onChange={[Function]}
              value={1}
            />
          </div>
        </li>
        <hr />
        <li
          className="sc-htoDjs bPGBDH"
        >
          <div
            className="sc-iwsKbI jQhHGA"
          >
            <span>
              Promo Code
            </span>
            <input
              className="sc-jzJRlG fsfCyH"
              onChange={[Function]}
              value=""
            />
          </div>
          <div
            className="sc-iwsKbI jQhHGA"
          >
            <span>
              Total
            </span>
            <div>
              A$99.99
            </div>
          </div>
        </li>
      </ul>
    </div>
  </main>,
  <footer
    className="sc-ifAKCX exVfAp"
  >
    <span>
      Built by 
       
       
      <a
        href="https://www.linkedin.com/in/paul-li-97381b3/"
        target="blank"
      >
        engineforce
      </a>
      , © 
      2019
    </span>
  </footer>,
]
`;

exports[`CartPage when render with no item in cart  1`] = `
Array [
  <header
    className="sc-bdVaJa llhFdL"
  >
    <h1
      className="sc-bwzfXH kabNnl"
    >
      <a
        aria-current="page"
        className=""
        href="/"
        onClick={[Function]}
        onMouseEnter={[Function]}
        style={Object {}}
      >
        Online Store
      </a>
    </h1>
  </header>,
  <nav
    className="sc-htpNat bRwrsF"
  >
    <div
      className="sc-bxivhb cLBQgY"
    >
      <a
        aria-current="page"
        className=""
        href="/"
        onClick={[Function]}
        onMouseEnter={[Function]}
        style={Object {}}
      >
        Home
      </a>
    </div>
    <div
      className="sc-bxivhb cLBQgY"
    >
      <a
        href="/products"
        onClick={[Function]}
        onMouseEnter={[Function]}
      >
        Products
      </a>
    </div>
    <div
      className="sc-bxivhb cLBQgY"
    >
      <a
        href="/cart"
        onClick={[Function]}
        onMouseEnter={[Function]}
      >
        Cart (
        0
        )
      </a>
    </div>
  </nav>,
  <main
    className="sc-EHOje ksOsdW"
  >
    <div
      className="sc-bZQynM iwmxdw"
    >
      <ul
        className="sc-gzVnrw lfdfns"
      >
        <li
          className="sc-htoDjs bPGBDH"
        >
          <div>
            <button
              className="sc-cSHVUG iFKOBr"
              onClick={[Function]}
            >
              EMPTY CART
            </button>
          </div>
        </li>
        <li
          className="sc-htoDjs sc-dnqmqq BIZtQ"
        >
          <div
            className="sc-iwsKbI sc-gZMcBi gdTgYj"
          >
            My Cart
          </div>
          <div
            className="sc-iwsKbI sc-gqjmRU fQHcrV"
          >
            Price
          </div>
          <div
            className="sc-iwsKbI sc-VigVT kGKNKu"
          >
            Qty
          </div>
        </li>
        <hr />
        <hr />
        <li
          className="sc-htoDjs bPGBDH"
        >
          <div
            className="sc-iwsKbI jQhHGA"
          >
            <span>
              Promo Code
            </span>
            <input
              className="sc-jzJRlG fsfCyH"
              onChange={[Function]}
              value=""
            />
          </div>
          <div
            className="sc-iwsKbI jQhHGA"
          >
            <span>
              Total
            </span>
            <div>
              A$0.00
            </div>
          </div>
        </li>
      </ul>
    </div>
  </main>,
  <footer
    className="sc-ifAKCX exVfAp"
  >
    <span>
      Built by 
       
       
      <a
        href="https://www.linkedin.com/in/paul-li-97381b3/"
        target="blank"
      >
        engineforce
      </a>
      , © 
      2019
    </span>
  </footer>,
]
`;
